### PR TITLE
Implement popup_during_fullscreen

### PR DESCRIPTION
**Pull request by RyanDwyer**
i3 docs: https://i3wm.org/docs/userguide.html#_popups_during_fullscreen_mode

~~Note that this currently doesn't work for xwayland. Xwayland popups are unmanaged, and for some reason they don't have the `NET_WM_WINDOW_TYPE_POPUP` atom when mapping, so I can't identify them from other unmanaged surfaces.~~

### CODE REVIEW

**Comment by RyanDwyer** [comment]
I've replaced the commit with one that does it using dialogs. From my commit message:

> This introduces a new `view_impl` function: `is_transient_for`. Similar to `container_has_ancestor` but works using the surface parents rather than the tree.
> 
> This patch modifies `view_is_visible`, `container_at` and so on to allow transient views to function normally when they're in front of a fullscreen view.

File: sway/desktop/xdg_shell.c [hunk 1]
```diff
@@ -385,6 +401,18 @@ static void handle_map(struct wl_listener *listener, void *data) {
 		view_update_csd_from_client(view, csd);
 	}
 
+	if (config->popup_during_fullscreen == POPUP_LEAVE &&
```
Comment (emersion) [inline sway/desktop/xdg_shell.c]: Can't this be done in a generic way with `is_transient_for`?

File: sway/desktop/xdg_shell.c [hunk 2]
```diff
@@ -385,6 +401,18 @@ static void handle_map(struct wl_listener *listener, void *data) {
...
+			view->container->workspace &&
+			view->container->workspace->fullscreen &&
+			xdg_surface->toplevel->parent) {
+		struct wlr_xdg_surface *psurface = xdg_surface->toplevel->parent;
+		struct sway_xdg_shell_view *parent = psurface->data;
+		struct sway_view *sway_view = &parent->view;
+		if (sway_view->container && sway_view->container->is_fullscreen) {
```
[the remark this answers is not in the document]
Comment (RyanDwyer) [reply sway/desktop/xdg_shell.c]: Yes, but in your example, fullscreen would have been exited when the parent was mapped.

[the remark this answers is not in the document]
Comment (emersion) [reply sway/desktop/xdg_shell.c]: xdg-shell allows weird stuff like: root (fullscreen, mapped) → parent (unmapped) → child (mapped). This is equivalent to: root (fullscreen, mapped) → child (mapped).

[the remark this answers is not in the document]
Comment (RyanDwyer) [reply sway/desktop/xdg_shell.c]: Amended.

File: sway/desktop/xdg_shell.c [hunk 3]
```diff
@@ -385,6 +401,17 @@ static void handle_map(struct wl_listener *listener, void *data) {
 		view_update_csd_from_client(view, csd);
 	}
 
+	if (config->popup_during_fullscreen == POPUP_LEAVE &&
```
Comment (emersion) [inline sway/desktop/xdg_shell.c]: Now that this part is generic, is it possible not to duplicate it and move it into `view_map`?

**Comment by RedSoxFan** [comment]
Thanks!

**Pull request merged**